#!/bin/bash

module load bbtools

# Use BBtools to demultiplex  files based on the headers
for file in *_R1.fastq.gz
  do
      bname=`basename $file _R1.fastq.gz`
      demuxbyname.sh in=$file in2=${bname}_R2.fastq.gz delimiter=colon column=10 out=${bname}_%_R1.fastq.gz out2=${bname}_%_R2.fastq.gz
  done

# cpy origional data to a new folder
mkdir original
mv L_Feed_R* original
mv 16S_Feed_R* original
mv L_Water_R* original
mv L_Gut_R* original
mv 16S_Water_R* original
mv 16S_Gut_R* original

# create the manifest
ls *.fastq.gz > filenames.txt
python filenames_2_manifest.py > manifest.txt
# 16S and LSU should be processed separatly, split them now
head -n1 manifest.txt > 16S_manifest.txt
head -n1 manifest.txt > LSU_manifest.txt
grep "LSU" manifest.txt >> LSU_manifest.txt
grep "16S" manifest.txt >> 16S_manifest.txt

source activate /project/gbru/gbru_fy18_candis/qiime2-2019.1

qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' \
                   --input-format PairedEndFastqManifestPhred33 \
                   --input-path 16S_manifest.txt \
                   --output-path 16S_demuxed.qza

qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' \
                   --input-format PairedEndFastqManifestPhred33 \
                   --input-path LSU_manifest.txt \
                   --output-path LSU_demuxed.qza

time qiime demux summarize \
  --i-data 16S_demuxed.qza \
  --o-visualization 16S_demuxed.qzv

time qiime demux summarize \
   --i-data LSU_demuxed.qza \
   --o-visualization LSU_demuxed.qzv

qiime dada2 denoise-paired --i-demultiplexed-seqs LSU_demuxed.qza \
  --p-trunc-len-f 220 \
  --p-trunc-len-r 150 \
  --p-n-threads 38 \
  --p-n-reads-learn 1000000 \
  --output-dir dada2_LSU \
  --verbose

qiime dada2 denoise-paired --i-demultiplexed-seqs 16S_demuxed.qza \
  --p-trunc-len-f 275 \
  --p-trunc-len-r 240 \
  --p-n-threads 38 \
  --p-n-reads-learn 1000000 \
  --output-dir dada2_16S \
  --verbose

# ad environemntal metadata
time qiime feature-table summarize \
  --i-table /dada2_LSU/table.qza \
  --o-visualization LSU_table-dada2.qzv \
  --m-sample-metadata-file LSU_metadata.txt

time qiime feature-table summarize \
  --i-table dada2_16S/table.qza \
  --o-visualization 16S_table-dada2.qzv \
  --m-sample-metadata-file 16S_metadata.txt

#align rep set sequences with mafft
time qiime alignment mafft \
  --i-sequences dada2_LSU/representative_sequences.qza \
  --o-alignment LSU_aligned-rep-seqs.qza \
  --p-n-threads 16

time qiime alignment mafft \
  --i-sequences dada2_16S/representative_sequences.qza \
  --o-alignment 16S_aligned-rep-seqs.qza \
  --p-n-threads 16

time qiime alignment mask \
  --i-alignment 16S_aligned-rep-seqs.qza \
  --o-masked-alignment 16S_masked-aligned-rep-seqs.qza

time qiime alignment mask \
  --i-alignment LSU_aligned-rep-seqs.qza \
  --o-masked-alignment LSU_masked-aligned-rep-seqs.qza


time qiime phylogeny fasttree \
  --i-alignment LSU_masked-aligned-rep-seqs.qza \
  --o-tree LSU_unrooted-tree.qza

time qiime phylogeny fasttree \
  --i-alignment 16S_masked-aligned-rep-seqs.qza \
  --o-tree 16S_unrooted-tree.qza

time qiime phylogeny midpoint-root \
  --i-tree 16S_unrooted-tree.qza \
  --o-rooted-tree 16S_rooted-tree.qza

time qiime phylogeny midpoint-root \
  --i-tree LSU_unrooted-tree.qza \
  --o-rooted-tree LSU_rooted-tree.qza


# download the  taxonomy from databases from Silva
mkdir taxonomy
mkdir taxonomy/SILVA_132_QIIME_release
cd taxonomy/SILVA_132_QIIME_release
wget https://www.arb-silva.de/fileadmin/silva_databases/qiime/Silva_132_release.zip
unzip Silva_132_release.zip
cd ../

#mkdir LSU_silva
#cd LSU_silva
# Silva 128 LSU data generated by Victor Carrillo -- Did not work only has bacteria
#https://forum.qiime2.org/t/large-23s-28s-lsu-subunit-ribosomal-rna-in-qiime-compatible-silva-database/3126/16
#wget https://www.dropbox.com/s/bcwo1qz0oci849k/LSU_cluster_99.zip?dl=0
#unzip LSU_cluster_99.zip?dl=0
#rm LSU_cluster_99.zip?dl=0

#format database
mkdir new_LSU
cd new_LSU
wget https://www.arb-silva.de/fileadmin/silva_databases/release_132/Exports/SILVA_132_LSURef_tax_silva_trunc.fasta.gz
gunzip SILVA_132_LSURef_tax_silva_trunc.fasta.gz
prep_silva_data.py --infile SILVA_132_LSURef_tax_silva_trunc.fasta \
  --taxafile LSU_taxonomy.txt \
  --outfasta LSU_centroids.fasta \
  --clusterid 0.99 \
  --threads 36

cd ../..




# import data, trim the databases to the region of interest, and train classiifer

## Silva_132_release SSU
qiime tools import \
  --type 'FeatureData[Sequence]' \
  --input-path taxonomy/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.fna \
  --output-path taxonomy/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.qza

qiime tools import \
  --type 'FeatureData[Taxonomy]' \
  --input-format HeaderlessTSVTaxonomyFormat \
  --input-path taxonomy/SILVA_132_QIIME_release/taxonomy/16S_only/99/majority_taxonomy_7_levels.txt \
  --output-path taxonomy/SILVA_132_QIIME_release/taxonomy/16S_only/99/majority_taxonomy_7_levels.qza

qiime feature-classifier extract-reads \
  --i-sequences taxonomy/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S.qza \
  --p-f-primer AGAGTTTGATCCTGGCTCAG \
  --p-r-primer ATTACCGCGGCTGCTGG \
  --p-min-length 300 \
  --p-max-length 600 \
  --o-reads taxonomy/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S_trimmed_27-534.qza

qiime feature-classifier fit-classifier-naive-bayes \
  --i-reference-reads taxonomy/SILVA_132_QIIME_release/rep_set/rep_set_16S_only/99/silva_132_99_16S_trimmed_27-534.qza \
  --i-reference-taxonomy taxonomy/SILVA_132_QIIME_release/taxonomy/16S_only/99/majority_taxonomy_7_levels.qza \
  --o-classifier 16S_classifier.qza


## Silva 132 LSU


qiime tools import \
  --type 'FeatureData[Sequence]' \
  --input-path taxonomy/new_LSU/LSU_centroids_clean.fasta \
  --output-path taxonomy/new_LSU/LSU_centroids.qza

qiime tools import \
  --type 'FeatureData[Taxonomy]' \
  --input-format HeaderlessTSVTaxonomyFormat \
  --input-path taxonomy/new_LSU/LSU_taxonomy.txt \
  --output-path taxonomy/new_LSU/LSU_taxonomy.qza

qiime feature-classifier extract-reads \
--i-sequences taxonomy/new_LSU/LSU_centroids.qza \
--p-f-primer AACKGCGAGTGAAGCRGBM \
--p-r-primer TCTTTCCCTCACGGTACTTG \
--p-min-length 100 \
--p-max-length 400 \
--verbose \
--o-reads taxonomy/new_LSU/LSU_taxonomy_trimmed_200-481.qza

qiime feature-classifier fit-classifier-naive-bayes \
--i-reference-reads taxonomy/new_LSU/LSU_taxonomy_trimmed_200-481.qza \
--i-reference-taxonomy taxonomy/new_LSU/LSU_taxonomy.qza \
--o-classifier LSU_classifier.qza


# CLassify taxonomically

## 16S
time qiime feature-classifier classify-sklearn \
--i-classifier  16S_classifier.qza \
--i-reads dada2_16S/representative_sequences.qza \
--o-classification 16S_taxonomy.qza \
--p-n-jobs 34 \
--verbose


time qiime feature-classifier classify-sklearn \
--i-classifier  LSU_classifier.qza \
--i-reads dada2_LSU/representative_sequences.qza \
--o-classification LSU_taxonomy.qza \
--p-n-jobs 34 \
--verbose

#plot samples
qiime taxa barplot \
  --i-table dada2_LSU/table.qza \
  --i-taxonomy LSU_taxonomy.qza \
  --m-metadata-file LSU_metadata.txt \
  --o-visualization LSU_taxa-bar-plots.qzv

  qiime taxa barplot \
    --i-table dada2_16S/table.qza \
    --i-taxonomy 16S_taxonomy.qza \
    --m-metadata-file 16S_metadata.txt \
    --o-visualization 16S_taxa-bar-plots.qzv

qiime diversity core-metrics-phylogenetic \
  --i-table dada2_16S/table.qza \
  --i-phylogeny 16S_rooted-tree.qza \
  --m-metadata-file 16S_metadata.txt \
  --p-n-jobs 30 \
  --output-dir 16S_diversity \
  --p-sampling-depth 5000

qiime diversity core-metrics-phylogenetic \
  --i-table dada2_LSU/table.qza \
  --i-phylogeny LSU_rooted-tree.qza \
  --m-metadata-file LSU_metadata.txt \
  --p-n-jobs 30 \
  --output-dir LSU_diversity \
  --p-sampling-depth 5000

# pip install deicode
# qiime dev refresh-cache
# this is the prefered ordination methods because the data is compositional count database
qiime deicode rpca \
   --i-table dada2_16S/table.qza \
   --p-min-feature-count 10 \
   --p-min-sample-count 500 \
   --o-biplot 16S_ordination.qza \
   --o-distance-matrix 16S_distance.qza

qiime emperor biplot \
   --i-biplot 16S_ordination.qza \
   --m-sample-metadata-file 16S_metadata.txt \
   --m-feature-metadata-file 16S_taxonomy.qza \
   --o-visualization 16S_biplot.qzv \
   --p-number-of-features 5

qiime deicode rpca \
  --i-table dada2_LSU/table.qza \
  --p-min-feature-count 10 \
  --p-min-sample-count 500 \
  --o-biplot LSU_ordination.qza \
  --o-distance-matrix LSU_distance.qza

qiime emperor biplot \
    --i-biplot LSU_ordination.qza \
    --m-sample-metadata-file LSU_metadata.txt \
    --m-feature-metadata-file LSU_taxonomy.qza \
    --o-visualization LSU_biplot.qzv \
    --p-number-of-features 5
